File: 113-path-sum-ii/113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    vector<vector<int>> numbers;
    
    
    int summer(vector<int> vec)
    {
        int sum = 0;
        
        for(int i=0; i<vec.size(); i++)
        {
            sum += vec[i];
        }
        
        return sum;
    }
    
    
    void nums(TreeNode* root, vector<int> vec)
    {
        if(!root->left && !root->right)
        {
            vec.push_back(root->val);
            
            numbers.push_back(vec);
            
            return;
        }
        
        
        vec.push_back(root->val);
        
        
        if(root->left)
            nums(root->left, vec);
        
        if(root->right)
           nums(root->right, vec);
        
    }
    
    
    vector<vector<int>> pathSum(TreeNode* root, int target) 
    {
        if(root == NULL)
            return numbers;
        
        vector<int> ans;
        
        nums(root, ans);
        
        vector<vector<int>> returner;
        
        for(int i=0; i<numbers.size(); i++)
        {
            int sum = summer(numbers[i]);
            
            if(sum == target)
                returner.push_back(numbers[i]);
        }
        
        return returner;
    }
};

File: 113-path-sum-ii/113. Path Sum II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    List<List<Integer>> ans = new ArrayList<>();
    
    public void recur(TreeNode root, int targetSum, List<Integer> temp, int currSum)
    {
        if(root.left == null && root.right == null)
        {
            if(currSum + root.val == targetSum)
            {
                // System.out.println(currSum + " " + root.val);
                
                temp.add(root.val);
                ans.add(new ArrayList<>(temp)); // copy
                temp.remove(temp.size() - 1);
            }
            return;
        }

        if(root.left != null)
        {
            temp.add(root.val);
            recur(root.left, targetSum, temp, currSum+root.val);
            temp.removeLast();
        }

        if(root.right != null)
        {
            temp.add(root.val);
            recur(root.right, targetSum, temp, currSum+root.val);
            temp.removeLast();
        }

        return;
    }

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {

        List<Integer> temp = new ArrayList<>();
        ans.clear();
        if(root == null)return ans;
        recur(root, targetSum, temp, 0);

        return ans;
    }
}